src/app/bulk-edit/bulk-edit-container/bulk-edit-container.component.ts

```diff
@@ -78,20 +79,38 @@ export class BulkEditContainerComponent implements OnInit {
   }
 
   cancel() {
-    this.editing
-      .confirmCancelAsync()
-      .pipe(filter((confirm) => !!confirm))
-      .subscribe(() => {
-        this.editing.stop();
-        this.stateHandler.GoPrevious();
-      });
+    const confirm$ =
+      this.currentStep === BulkEditStep.Editor && this.hasChanged
+        ? this.editing.confirmCancelAsync()
+        : of(true);
+
+    confirm$.pipe(filter((confirm) => !!confirm)).subscribe(() => {
+      this.editing.stop();
+      this.stateHandler.GoPrevious();
+    });
   }
 
   next() {
     this.currentStep = this.currentStep + 1;
+    this.hasChanged = false;
   }
 
   previous() {
-    this.currentStep = this.currentStep - 1;
+    const confirm$ = this.hasChanged
+      ? this.editing.confirmCancelAsync()
+      : of(true);
+
+    confirm$.pipe(filter((confirm) => !!confirm)).subscribe(() => {
+      this.currentStep = this.currentStep - 1;
+      this.hasChanged = false;
+    });
+  }
+
+  onChanged() {
+    this.hasChanged = true;
+  }
+
+  onSaved() {
+    this.hasChanged = false;
   }
```

Comment:
Not necessary to change, but wondering if observables could/should be used instead of these setters. Something like, create a `hasChangedEvent$` stream which listens to a combined stream of the changed/saved button click events in the child component and uses the latest emission to set the value of `hasChanged`. mainly I'm just curious if you have any thoughts about that as an alternative method to traditional event bindings. Certainly, this way appears simpler/easier to grok at a glance.